x/src/cargo.rs

```diff
@@ -115,9 +128,15 @@ impl Cargo {
         self
     }
 
+    /// no fail fast will prevent propigation of errors.
     pub fn run(&mut self) -> Result<()> {
         self.inner.stdout(Stdio::inherit()).stderr(Stdio::inherit());
-        self.do_run(true).map(|_| ())
+        let output = self.do_run(true).map(|_| ());
+        if self.no_fail_fast {
+            Ok(())
+        } else {
+            output
+        }
```

Comment:
My recommendation is to not hide this internally as it may make things difficult to understand, especially since when running with --no-fail-fast the exit code will still be non-zero and we want to reflect that in the return value here. Instead you should have logic in an upper layer which inspects the result from this call and figure out what to do.

Reply:
Agree.   Been thinking about how to structure this.   Seems to me Cargo should only be a "command factory", translate external commands in to internal "commands".   Those commands should not be executed by the command factory, but only be data.   That data can be passed to an CommandExecutor.   That executor could have multiple observers/decorators around commands, which could perform logging, tracking timing,  perform retries, choose to ignore the commands and simply log out the data.   I'll check to see what exists in crates around this.  The lack of these necessary components/separation between them may be the cause of some friction/cognitive dissonance we're both feeling around this now.   But, again I'd prefer to see what the feature set will need before I'd want to refactor fully to these patterns, as this is just a guess based on what I've been thinking through over the weekend.